homeassistant/components/zha/config_flow.py

```diff
@@ -53,48 +183,261 @@ async def async_step_user(self, user_input=None):
+                    network_info=app.state.network_info,
+                    node_info=app.state.node_info,
+                )
+
+        strategies = []
+
+        # Check if we have any automatic backups *and* if the backups differ from
+        # the current radio settings (since restoring would be redundant)
+        if self._backups and any(
+            not backup.is_compatible_with(self._current_settings)
+            for backup in self._backups
+        ):
+            strategies.append(FORMATION_RESTORE_AUTOMATIC_BACKUP)
+
+        if self._current_settings is not None:
+            strategies.append(FORMATION_REUSE_SETTINGS)
+
+        strategies.append(FORMATION_RESTORE_MANUAL_BACKUP)
+        strategies.append(FORMATION_FORM_NEW_NETWORK)
+
+        # Pick a default. Forming a new network will always be an option but we would
+        # prefer to suggest something non-destructive if it's available.
+        suggested_strategy = [
+            strategy
+            for strategy in strategies
+            if strategy
+            in (
+                FORMATION_RESTORE_AUTOMATIC_BACKUP,
+                FORMATION_REUSE_SETTINGS,
+                FORMATION_FORM_NEW_NETWORK,
+            )
+        ][0]
+
+        schema = {
+            vol.Required(FORMATION_STRATEGY, default=suggested_strategy): vol.In(
+                strategies
+            ),
+        }
+
+        return self.async_show_form(
+            step_id="choose_formation_strategy",
+            data_schema=vol.Schema(schema),
+        )
+
+    def _parse_uploaded_backup(
+        self, uploaded_file_id: str
+    ) -> zigpy.backups.NetworkBackup:
+        """Read and parse an uploaded backup JSON file."""
+        with process_uploaded_file(self.hass, uploaded_file_id) as file_path:
+            contents = file_path.read_text()
+
+        return zigpy.backups.NetworkBackup.from_dict(json.loads(contents))
+
+    async def async_step_upload_manual_backup(self, user_input=None):
+        """Upload and restore a coordinator backup JSON file."""
+        errors = {}
+
+        if user_input is not None:
+            try:
+                backup = await self.hass.async_add_executor_job(
+                    self._parse_uploaded_backup, user_input[UPLOADED_BACKUP_FILE]
+                )
+            except ValueError:
+                errors["base"] = "invalid_backup_json"
+            else:
+                if user_input.get(OVERWRITE_COORDINATOR_IEEE):
+                    backup = self._allow_overwrite_ezsp_ieee(backup)
+
+                async with self._connect_zigpy_app() as app:
+                    await app.backups.restore_backup(backup)
+
+                return await self._async_create_radio_entity()
+
+        data_schema = {
+            vol.Required(UPLOADED_BACKUP_FILE): FileSelector(
+                FileSelectorConfig(accept=".json,application/json")
+            )
+        }
+
+        if self._radio_type == RadioType.ezsp and (
```

Comment:
Should we split this out into a confirm step and only show when we detect that the backup has a different IEEE ?

Reply:
That works. It's an extra step but it does make things clearer and can be used for both automatic and manual backup restoration.